### PR TITLE
Made changes to routes, controllers, and views

### PR DESCRIPTION
**Pull request by barobba**
Hi Peter, there are a LOT checked in here, but it's based on our conversation about cleaning up the routes, controllers, and views.

### CODE REVIEW

File: app/Http/Controllers/Group/Workspace/TopicController.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+use App\Models\Bookmark;
+use App\Models\Group;
+use App\Models\LegiScan\Session;
+use Carbon\Carbon;
+use Illuminate\Database\Eloquent\Builder;
+use Illuminate\Http\Request;
+
+class TopicController extends Controller
+{
+    public function show(Group $group, Group $workspace, Group $topic)
```
Comment (PeterDKC) [inline app/Http/Controllers/Group/Workspace/TopicController.php:15]: Everything that leads up to this controller method needs to be rethought from scratch. This should scream "I haven't thought through my database structure and Model design" in 10-foot-tall neon letters to you. No Controller method should have contents like this.

Ideal Controller method:

- Accept & validate a request
- Perform simple logic or defer to complex objects
- Return a response

This Controller method:

- Figure out masses of confusing nested SQL to build an entire world from scratch and then serve up a hot mess back to the view to figure out with every Model in the database all at once for pretty much no reason.

File: app/Http/Controllers/Group/WorkspaceController.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+use App\Models\Group;
+use App\Models\LegiScan\Session;
+use Carbon\Carbon;
+use Illuminate\Database\Eloquent\Builder;
+use Illuminate\Http\Request;
+
+class WorkspaceController extends Controller
+{
+    public function show(Group $group, Group $workspace)
+    {
```
Comment (PeterDKC) [inline app/Http/Controllers/Group/WorkspaceController.php:16]: copy-paste above comments

**Review by PeterDKC — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/Group/AboutController.php [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+use App\Http\Controllers\Controller;
+use App\Models\Group;
+use Illuminate\Http\Request;
+
+class AboutController extends Controller
+{
+    public function show(Group $group)
+    {
+        return view('groups.about', compact('group'));
```
[the remark this answers is not in the document]
Comment (barobba) [reply app/Http/Controllers/Group/AboutController.php:13]: So, views should also have the method at the end? In this case it would be `groups.about.show`?

In that case, they would only implement: `index`, `show`, `create`, `edit`?

These wouldn't be used for views: `store`, `update`, `delete`?

Did I get this right?

File: app/Http/Controllers/Group/AboutController.php [hunk 4]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+namespace App\Http\Controllers\Group;
+
+use App\Http\Controllers\Controller;
+use App\Models\Group;
+use Illuminate\Http\Request;
```
[the remark this answers is not in the document]
Comment (barobba) [reply app/Http/Controllers/Group/AboutController.php:7]: These are used though? The only one that isn't is the `Request` class.

File: app/Http/Controllers/Group/AboutController.php [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+class AboutController extends Controller
```
[the remark this answers is not in the document]
Comment (barobba) [reply app/Http/Controllers/Group/AboutController.php:9]: Ok, what should it be named?

File: app/Http/Controllers/Group/NavigateStateController.php [hunk 6]
```diff
@@ -1,11 +1,12 @@
...
 
-namespace App\Http\Controllers;
+namespace App\Http\Controllers\Grouop;
 
+use App\Http\Controllers\Controller;
 use App\Models\Group;
 use Illuminate\Http\Request;
 
-class GroupNavigateStateController extends Controller
+class NavigateStateController extends Controller
```
[the remark this answers is not in the document]
Comment (barobba) [reply app/Http/Controllers/Group/NavigateStateController.php:9]: I'm confused, are you saying that I would simply use the session in the controller and in the view?

I thought it would be more fluent to have the controller use `$group->chooseYear($year)` and the view would use {{ $group->chosenYear() }}, instead of requiring session logic, which doesn't mean anything. Where do I put the function that creates the session key?

What am I missing here?

File: app/Http/Controllers/Group/Workspace/Topic/BillSearchController.php [hunk 7]
```diff
@@ -1,17 +1,16 @@
...
 use App\Models\LegiScan\Session;
-use App\Models\LegiScan\State;
 use Carbon\Carbon;
 use Illuminate\Http\Request;
 
-class GroupBillSearchController extends Controller
+class BillSearchController extends Controller
 {
-    public function show(Group $group, State $state, Session $session)
+    public function show(Group $group, Group $workspace, Group $topic)
```
[the remark this answers is not in the document]
Comment (barobba) [reply app/Http/Controllers/Group/Workspace/Topic/BillSearchController.php:13]: Can we go over this for the next GitHub issue/pull request?

File: app/Http/Controllers/Group/Workspace/TopicController.php (hunk 1 above)
Comment (barobba) [reply app/Http/Controllers/Group/Workspace/TopicController.php:15]: The main issue you have is three variables with the same class?

File: app/Http/Controllers/Group/AboutController.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (barobba) [reply app/Http/Controllers/Group/AboutController.php:13]: Do I also need `dashboard.show` and `login.show`?

File: app/Http/Controllers/DashboardController.php [hunk 8]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+namespace App\Http\Controllers;
+
+// use App\Models\LegiScan\State;
```
[the remark this answers is not in the document]
Comment (PeterDKC) [reply app/Http/Controllers/DashboardController.php]: Delete the commented line.

File: app/Http/Controllers/DashboardController.php [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+use App\Http\Controllers\Controller;
+use Illuminate\Support\Facades\Auth;
+
+class DashboardController extends Controller
+{
+    public function index()
+    {
+        $user = Auth::user();
+
+        return view('dashboard', compact('user'));
```
[the remark this answers is not in the document]
Comment (PeterDKC) [reply app/Http/Controllers/DashboardController.php:14]: There's no reason to do it like this. No reason to force yourself to change everything or nothing in the codebase at one time. If you're looking at it and you have a new way of doing it just change it. The way you're approaching it does not scale with the size of the repo.

File: app/Http/Controllers/Group/AboutController.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (PeterDKC) [reply app/Http/Controllers/Group/AboutController.php]: I think the point on this one in particular - discounting special cases like `dashboard/login/logout/home/welcome`, is that either this is the `groups.show` page and the other thing is the `groups.bills.index` page, or this is the `groups.members.index` page and the other is the `groups.show` page. But either way one or both of your routes is miscategorized.

File: app/Http/Controllers/Group/AboutController.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (PeterDKC) [reply app/Http/Controllers/Group/AboutController.php]: The comment is on the `Request` line, therefore I'm referring to the `Request` line, not all the use statements.

File: app/Http/Controllers/Group/AboutController.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (PeterDKC) [reply app/Http/Controllers/Group/AboutController.php:8]: What class or model is primarily being indexed / shown / updated / created / deleted?

File: app/Http/Controllers/Group/NavigateStateController.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (PeterDKC) [reply app/Http/Controllers/Group/NavigateStateController.php:9]: I'm not saying you need to do this in this PR, but for the sake of all that is holy in the world, please plan to delete absolutely everything about what you're doing here and never speak of it again... 😢

File: app/Http/Controllers/Group/Workspace/TopicController.php (hunk 1 above)
Comment (PeterDKC) [reply app/Http/Controllers/Group/Workspace/TopicController.php:14]: No, it's the entire contents of the controller method.

File: app/Http/Controllers/Group/AboutController.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (barobba) [reply app/Http/Controllers/Group/AboutController.php]: I don't understand what you are saying here.

File: app/Http/Controllers/Group/AboutController.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (barobba) [reply app/Http/Controllers/Group/AboutController.php:8]: This is not an API, these are for Web pages. I might want 20 pages that show the group in different ways. According to the CRUDy way, wouldn't each of the 20 pages get its own controller, even if they're for the same model?

File: app/Http/Controllers/Group/NavigateStateController.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (barobba) [reply app/Http/Controllers/Group/NavigateStateController.php:9]: What this is doing is letting the user save the preferred year and state, for the lifetime of the session, for each group they are working in. 

The reason for the key is not a route. It's just a unique name for the key in the user's session.

File: app/Http/Controllers/Group/AboutController.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (PeterDKC) [reply app/Http/Controllers/Group/AboutController.php]: In a nutshell, `about` is not a thing in a CRUD app. (Which, pretty much all non-pamphlet Laravel apps are)

So how can you reorganize the routes and controller names to delete that word and use CRUD terms and organization instead?

It's a page that shows properties of a single group. It's constituent members, name, etc etc. Generally, that's a `show` page. 

Your other current `.show` page is really showing a list of things. What are the things it's showing a list of? That's probably a `groups.<that-model-name-plural>.index` route.

As I mentioned initially, your route organization does not make any sense and the way you're trying to organize is leading you into really unhealthy patterns like returning 3 views conditionally from a `show()` method and putting CRUD route keys into the `session()`. It's all connected.

If you can figure out this one piece, it may untangle other aspects for you to be able to reorganize into a sane overall principle and eliminate the unhealthy aspects of the current approach.

File: app/Http/Controllers/Group/AboutController.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (PeterDKC) [reply app/Http/Controllers/Group/AboutController.php:8]: It doesn't matter. It's not just APIs that are organized this way. Every grown up web application, .NET, Java, Flask, Django, Laravel, Symfony, ... literally name a language / web framework, and grown up software engineers organize their application this way. CRUDful organization is the default for the non-CMS web as of 20 years ago, and you should adhere to it always in non-CMS applications.

> I might want 20 pages that show the group in different ways

No. That's not how you do this. Or rather that's not how people who construct well built applications that scale do this.

File: app/Http/Controllers/DashboardController.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (barobba) [reply app/Http/Controllers/DashboardController.php]: I converted every instance over to your method.

File: app/Http/Controllers/Group/AboutController.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (barobba) [reply app/Http/Controllers/Group/AboutController.php]: Ok, I changed `groups.about.show` to `groups.members.index`.

File: app/Http/Controllers/Group/AboutController.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (barobba) [reply app/Http/Controllers/Group/AboutController.php]: Ok, fixed!

File: app/Http/Controllers/Group/Workspace/TopicController.php (hunk 1 above)
Comment (barobba) [reply app/Http/Controllers/Group/Workspace/TopicController.php]: Fixed!

File: routes/web.php [hunk 10]
```diff
@@ -23,21 +22,38 @@
...
-    Route::get('group/{group}/search', [GroupBillSearchController::class, 'show'])->name('group.bill-search.show');
-
-    Route::put('group/{group}/navigate-state', [GroupNavigateStateController::class, 'update'])->name('group.navigate.state.update');
-    Route::put('group/{group}/navigate-year', [GroupNavigateYearController::class, 'update'])->name('group.navigate.year.update');
-
+// Routes requiring authentication
+Route::middleware('auth')->group(function()
+{
+    // Dashboard
+    Route::get('/dashboard', [DashboardController::class, 'index'])->name('dashboard');
```
Comment (PeterDKC) [inline routes/web.php:33]: `get('dashboard',...` the leading `/` doesn't do anything

File: app/Http/Controllers/Group/Workspace/Topic/BillSearchController.php [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+class BillSearchController extends Controller
+{
+    public function show(Group $group, Workspace $workspace, Topic $topic)
+    {
+        $sessionYears = Session::validYears();
+
+        return view('groups.workspaces.topics.bill-search')
+            ->withGroup($group)
+            ->withWorkspace($workspace)
+            ->withTopic($topic)
```
Comment (PeterDKC) [inline app/Http/Controllers/Group/Workspace/Topic/BillSearchController.php:21]: Technically you can just pass `$topic` and derive everything else from there, right?

```
$topic
$topic->workspace
$topic->workspace->group
```

You _can_ do it this way, but I would advise in general just passing the primary model you require.

File: app/Http/Controllers/Group/WorkspaceController.php [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+use App\Models\Bookmark;
+use App\Models\Group;
+use App\Models\LegiScan\Session;
+use App\Models\Workspace;
+
+class WorkspaceController extends Controller
+{
+    public function show(Group $group, Workspace $workspace)
+    {
+        $sessionYears = Session::validYears();
```
Comment (PeterDKC) [inline app/Http/Controllers/Group/WorkspaceController.php:15]: I get why you did this. However. This is now a testing blockade, because the method is `static`. This dependency cannot be mocked and registered in the container. It cannot be avoided.

I _think_ you can get away with it here, sort of, but the much better think to do is to make a third class that has the single responsibility of assembling this query, and then pick one of:

- `app(ValidSessions::class)->getValidYears()`
- Create a Facade to wrap that class in and call `ValidSessions::validYears()`

Either way, because you have an instantiable class with a non-static method, you can mock and insert it into the service container and avoid a blockade.

Again, you're probably fine for the moment and can refactor this later. However. More troubling to my OCD:

- Classes & Properties are Nouns
- Methods are Verbs
- "Valid Years" is a Noun. Please change it to be some kind of verb.

File: app/Http/Controllers/Group/WorkspaceController.php [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+        $bookmarks = Bookmark::query()
+                        ->perWorkspace($workspace)
```
Comment (PeterDKC) [inline app/Http/Controllers/Group/WorkspaceController.php:18]: As long as you're going to all the trouble to get session years out of here: I would make yourself a ticket to create 3 methods:

```
group->findBookmarks();
workspace->findBookmarks();
topic->findBookmarks();
```

since the only reason you're doing it all inline now is because you need to know `per<Model>(<model>)`. If you define it on all 3 models, that problem is solved.

File: app/Models/Group.php [hunk 14]
```diff
@@ -180,19 +106,14 @@ public function chooseYear($year)
...
     {
-        $state_abbr = session($this->chosenStateKey(), $this->root()->state_abbr);
-
-        if ($state_abbr)
-        {
-            return State::where('state_abbr', $state_abbr)->first();
-        }
-
-        return null;
+        $state_abbr = session($this->chosenStateKey(), $this->state_abbr);
```
Comment (PeterDKC) [inline app/Models/Group.php:113]:
```
$variableName
ClassTraitOrInterfaceName
$className->property_name
$className->methodName()
global_function_call()
```

Should be

```
$stateAbbr
```

I believe if you run `composer cs && composer phpmd` it will tell you about these things and enforce the standards for you.

File: app/Models/LegiScan/Session.php [hunk 15]
```diff
@@ -70,4 +71,21 @@ public function getTaglineAttribute()
     }
 
     #endregion
+
+    #region Static methods
+
+    public static function validYears()
```
Comment (PeterDKC) [inline app/Models/LegiScan/Session.php:77]: `static` 😨

**Review by PeterDKC — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/Group/Workspace/Topic/BillSearchController.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (barobba) [reply app/Http/Controllers/Group/Workspace/Topic/BillSearchController.php]: I got this fixed.

File: app/Http/Controllers/Group/Workspace/Topic/BillSearchController.php (hunk 11 above)
Comment (barobba) [reply app/Http/Controllers/Group/Workspace/Topic/BillSearchController.php:21]: I can do that, but in this case, why wouldn't my route just be `topic/{topic}` without being nested under group or workspace?

Comment (PeterDKC) [reply app/Http/Controllers/Group/Workspace/Topic/BillSearchController.php:21]: No. I get why you would think that, but that's only because you're only thinking about this method. There's 7 CRUD methods per resource level, and any of the others may require one or both of the other models in the chain in order to function.

It's just that for this one method, you technically only need the topic in order to satisfy the view.

File: app/Http/Controllers/Group/WorkspaceController.php (hunk 12 above)
Comment (barobba) [reply app/Http/Controllers/Group/WorkspaceController.php:15]: What about `SessionHelper::getValidYears()`? I think I remember you are not a fan of the word "helper".

File: app/Http/Controllers/Group/WorkspaceController.php (hunk 13 above)
Comment (barobba) [reply app/Http/Controllers/Group/WorkspaceController.php:18]: Done

File: app/Models/Group.php (hunk 14 above)
Comment (barobba) [reply app/Models/Group.php:113]: Got it... fixed!

File: app/Models/LegiScan/Session.php (hunk 15 above)
Comment (barobba) [reply app/Models/LegiScan/Session.php:77]: Fixed

File: app/Http/Controllers/Group/WorkspaceController.php (hunk 12 above)
Comment (barobba) [reply app/Http/Controllers/Group/WorkspaceController.php:15]: Fixed

File: app/Http/Controllers/Group/Workspace/Topic/BillSearchController.php (hunk 11 above)
Comment (barobba) [reply app/Http/Controllers/Group/Workspace/Topic/BillSearchController.php:21]: Ok. For now, I will keep it in... because I'm lazy. I can always refactor it out later, if that's ok.

File: app/Http/Livewire/BillTable.php [hunk 16]
```diff
@@ -40,12 +41,12 @@ public function query()
     {
         return Bill::query()
             ->when(
-                $this->scope->chosenState(),
-                fn ($query) => $query->where('state_id', $this->scope->chosenState()->id)
+                $this->group->chosenState(),
+                fn ($query) => $query->where('state_id', $this->group->chosenState()->id)
```
Comment (PeterDKC) [inline app/Http/Livewire/BillTable.php:45]: Don't fix any of this now, just be aware that

- This method does not belong on this class, it should be on `Bill`
- The use of `fn () => ;` throughout severely hurts the readability & clarity of this whole class
- The following makes my OCD go wild:
```
            })
            ->orderByDesc('created')
            ;
```
😱 
- Again, the session use here needs to be ripped out wholesale, so all this can be re-examined as part of that effort.

Just a mental nugget to consider.

File: app/Models/Bookmark.php [hunk 17]
```diff
@@ -36,14 +45,54 @@ public function bookmarkable()
 
     #endregion
 
-    #region Scopes
-
-    public function scopeWhereBills(Builder $query)
+    public function scopePerGroup(Builder $query, Group $group)
```
Comment (PeterDKC) [inline app/Models/Bookmark.php:48]: Running `composer cs-fix` should fix all of the whitespace formatting issues in the following ~50 lines

File: app/Models/User.php [hunk 18]
```diff
@@ -66,10 +66,10 @@ function groups_owned()
 
     public function getFullNameAttribute()
     {
-        $name_parts = array_filter([$this->first_name, $this->last_name]);
-        if ($name_parts)
+        $nameParts = array_filter([$this->first_name, $this->last_name]);
```
Comment (PeterDKC) [inline app/Models/User.php:69]: Why are you filtering these??.... Are you not requiring both at the User creation level??... 🤔 I can't figure out why you would do this...

File: bootstrap/helpers.php [hunk 19]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+use App\Helpers\LegislativeSessionHelper;
+
+function legislativeSessionHelper()
```
Comment (PeterDKC) [inline bootstrap/helpers.php:5]: .... ....       .............. Why????.....

File: config/app.php [hunk 20]
```diff
@@ -210,7 +215,9 @@
     */
 
     'aliases' => Facade::defaultAliases()->merge([
-        // 'ExampleClass' => App\Example\ExampleClass::class,
+
+        'LegislativeSessionHelper' => App\Helpers\Facades\LegislativeSessionHelperFacade::class,
```
Comment (PeterDKC) [inline config/app.php:219]: Rename these per comments above.

File: database/factories/LegiScan/BodyFactory.php [hunk 21]
```diff
@@ -16,18 +16,18 @@ public function definition()
     {
         static $increment = 1;
 
-        $body_name = ucfirst($this->faker->text(80));
-        $role_name = ucfirst($this->faker->text(80));
+        $bodyName = ucfirst($this->faker->text(80));
```
Comment (PeterDKC) [inline database/factories/LegiScan/BodyFactory.php:19]: What's important about `80`?... Why can't these just be `'body_name' => $this->faker->sentence(),`?... Not obvious. Is there a good reason or you just picked a random number?...

**Review by PeterDKC — CHANGES_REQUESTED** [review]

File: app/Helpers/LegislativeSessionHelper.php [hunk 22]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+namespace App\Helpers;
+
+use App\Models\LegiScan\Session;
+use Carbon\Carbon;
+
+class LegislativeSessionHelper
+{
+    public function validYears()
```
[the remark this answers is not in the document]
Comment (barobba) [reply app/Helpers/LegislativeSessionHelper.php:10]: Great! This really gets me thinking about naming things. 

I didn't mean to ignore your part about verbs, but I thought if I give it a first try with placing the code where it needs to go, then we could talk more about it.

... working on it now!

File: app/Providers/LegislativeSessionHelperProvider.php [hunk 23]
```diff
@@ -0,0 +1,31 @@
+<?php
+
+namespace App\Providers;
+
+use App\Helpers\LegislativeSessionHelper;
+use Illuminate\Support\ServiceProvider;
+
+class LegislativeSessionHelperProvider extends ServiceProvider
```
[the remark this answers is not in the document]
Comment (barobba) [reply app/Providers/LegislativeSessionHelperProvider.php:8]: Yes, I was trying both ways you suggested, and also Googling how to do it. 

So, I kind of thought these two were overkill.

I like the `app(MyClass::class)->doSomething()` approach. It's very easy to use.

[the remark this answers is not in the document]
Comment (barobba) [reply app/Providers/LegislativeSessionHelperProvider.php:8]: Working on removing it.

File: app/Helpers/Facades/LegislativeSessionHelperFacade.php [hunk 24]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace App\Helpers\Facades;
+
+use Illuminate\Support\Facades\Facade;
+
+class LegislativeSessionHelperFacade extends Facade
+{
+    protected static function getFacadeAccessor()
+    {
+        return 'legislativeSessionHelper';
```
[the remark this answers is not in the document]
Comment (barobba) [reply app/Helpers/Facades/LegislativeSessionHelperFacade.php:11]: Working on removing it.

File: app/Models/User.php (hunk 18 above)
Comment (barobba) [reply app/Models/User.php:69]: They are both required. I guess it's a bit extra here.

Are you saying, it can just be `$this->first_name.' '.$this->last_name`, and keep it simple?

File: config/app.php [hunk 25]
```diff
@@ -196,6 +196,11 @@
         App\Providers\EventServiceProvider::class,
         App\Providers\RouteServiceProvider::class,
 
+        /*
+         * Custom providers
+         */
+        App\Providers\LegislativeSessionHelperProvider::class,
```
[the remark this answers is not in the document]
Comment (barobba) [reply config/app.php:202]: Will do

File: bootstrap/helpers.php (hunk 19 above)
Comment (barobba) [reply bootstrap/helpers.php:5]: Lol, to bee more like how Laravel does it with `request('key')` or `auth()->user()`.

Comment (barobba) [reply bootstrap/helpers.php:5]: If it's not Peter-ok'd then... removing it now.

File: config/app.php (hunk 20 above)
Comment (barobba) [reply config/app.php:219]: Will do

File: database/factories/LegiScan/BodyFactory.php (hunk 21 above)
Comment (barobba) [reply database/factories/LegiScan/BodyFactory.php:19]: Body name is like: Senate, House of Representatives, House of Delegates, etc. ... I guess just to keep it short? Technically, it can be any string.

Comment (barobba) [reply database/factories/LegiScan/BodyFactory.php:19]: I will just use the `sentence()` method you have. Working on it.

File: app/Models/User.php (hunk 18 above)
Comment (PeterDKC) [reply app/Models/User.php:69]: Yeah. The array filter is only needed to filter out `null` value'd keys.

```
array_filter([null, 42]) // [42]
```

Just concatenate & return imho

File: bootstrap/helpers.php (hunk 19 above)
Comment (PeterDKC) [reply bootstrap/helpers.php:5]: But those are there for specific reasons, namely so that they can be used in situations where the app doesn't already exist and Facades are not a thing. When would you ever face that with this method?

File: app/Helpers/Models/LegiScan/SessionYear.php [hunk 26]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace App\Helpers\Models\LegiScan;
```
Comment (PeterDKC) [inline app/Helpers/Models/LegiScan/SessionYear.php:3]: ... Why is this namespace so complicated?? ... What value is that adding?

Also, I never said to NOT make a Facade. I just said to not make a Facade the hardest way you can make a facade...

Can this just be `App\Helpers\SessionYear`?

Can there be an `App\Facades\SessionYear` that just says

```
return app(SessionYear::class);
```

And then everywhere else we just say

```
use App\Facades\SessionsYear
...
SesssionYear::getAll()
```

or

```
{ \App\FacadesSessionYear::getAll()}
``

in views?...

File: tests/Unit/Models/UserTest.php [hunk 27]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace Tests\Unit\Models;
+
+use App\Models\User;
+use Tests\TestCase;
+
+class UserTest extends TestCase
+{
+    /** @test */
+    public function example()
```
Comment (PeterDKC) [inline tests/Unit/Models/UserTest.php:11]: example?

File: tests/Unit/Models/StateTest.php [hunk 28]
```diff
@@ -1,33 +0,0 @@
-<?php
-
-namespace Tests\Unit\Models;
```
Comment (PeterDKC) [inline tests/Unit/Models/StateTest.php:3]: Why is this deleted?...

File: tests/Unit/Helpers/SessionYearTest.php [hunk 29]
```diff
@@ -0,0 +1,33 @@
...
+        Session::factory()->create([
+            'year_start' => 3,
+            'year_end' => 4,
+        ]);
+
+        $years = app(SessionYear::class)->getAll();
+
+        $this->assertTrue($years->duplicates()->isEmpty());
+
+        $this->assertTrue($years->diff([1, 2, 3, 4])->isEmpty());
```
Comment (PeterDKC) [inline tests/Unit/Helpers/SessionYearTest.php:31]: Just assert that the collection matches the array expectations

```
$this->assertEquals(
  [1, 2, 3, 4],
  $myCollection->toArray()
);
```

🤷 Way easier

**Review by PeterDKC — CHANGES_REQUESTED** [review]

File: app/Helpers/Models/LegiScan/SessionYear.php [hunk 30]
```diff
@@ -0,0 +1,16 @@
...
+
+use App\Models\LegiScan\Session;
+
+class SessionYear
+{
+    public function getAll()
+    {
+        // Assumes session is no more than 2 years (e.g. no third year between the start and end years)
```
[the remark this answers is not in the document]
Comment (barobba) [reply app/Helpers/Models/LegiScan/SessionYear.php:11]: Sure

File: app/Helpers/Models/LegiScan/SessionYear.php [hunk 31]
```diff
@@ -0,0 +1,16 @@
...
+        return collect(Session::select('year_start', 'year_end')->get()->toArray())
```
[the remark this answers is not in the document]
Comment (barobba) [reply app/Helpers/Models/LegiScan/SessionYear.php:12]: The `$models->flatten()` doesn't work on models, and the result of `Session::select(...)->get()->toArray()` is no longer a collection so I wrap it in a collection.

I updated the nesting so it's more obvious.

File: app/Http/Controllers/Group/HearingController.php [hunk 32]
```diff
@@ -0,0 +1,38 @@
+<?php
+
+namespace App\Http\Controllers\Group;
```
[the remark this answers is not in the document]
Comment (barobba) [reply app/Http/Controllers/Group/HearingController.php:3]: Fixed!

File: app/Models/LegiScan/BillHistory.php [hunk 33]
```diff
@@ -2,14 +2,108 @@
...
+    {
+        return $query
+            ->whereHas('bill.session', function(Builder $query) use($year){
+                $query->where('year_start', $year)
+                      ->orWhere('year_end', $year);
+            });
+    }
+    public function scopeWhereIsHearing(Builder $query)
+    {
+        // Only working for NE, since we don't know how other state's hearings are presented in the history
```
[the remark this answers is not in the document]
Comment (barobba) [reply app/Models/LegiScan/BillHistory.php:69]: Ok, will do. 

I feel like scopes that represent conditional filters should have a conditional-sounding name. 
What about this name: `scopeWhereIsNebraskaHearing`?

File: app/Models/LegiScan/BillHistory.php [hunk 34]
```diff
@@ -2,14 +2,108 @@
...
+                $query->where('state_abbr', 'NE');
+            })
+            ->where('history_action', 'like', '%hearing%');
+    }
+
+    #endregion
+
+    #region Attributes
+
+    public function getStepAttribute()
```
[the remark this answers is not in the document]
Comment (barobba) [reply app/Models/LegiScan/BillHistory.php:81]: So my thought here is if we do swap out the models for our own, then the naming would be more consistent. Like less to refactor.

File: app/Models/LegiScan/Progress.php [hunk 35]
```diff
@@ -12,4 +12,9 @@ class Progress extends Model
     protected $table = 'ls_progress';
 
     protected $primaryKey = 'progress_event_id';
+
+    public function getDescriptionAttribute()
```
[the remark this answers is not in the document]
Comment (barobba) [reply app/Models/LegiScan/Progress.php:16]: But `progress_desc` is a horrible name.

[the remark this answers is not in the document]
Comment (barobba) [reply app/Models/LegiScan/Progress.php:16]: I'd rather paint the turd, until the day comes when I can get rid of it.

File: app/Helpers/Models/LegiScan/SessionYear.php (hunk 26 above)
Comment (barobba) [reply app/Helpers/Models/LegiScan/SessionYear.php:3]: I was making the helper keep a consistent nesting as everything else.

I might need to do a quick Zoom for this one.

Comment (barobba) [reply app/Helpers/Models/LegiScan/SessionYear.php:3]: ... or a Slack huddle.

File: tests/Unit/Models/UserTest.php (hunk 27 above)
Comment (barobba) [reply tests/Unit/Models/UserTest.php:11]: Whoops! Lol, fixing that now.

File: tests/Unit/Models/StateTest.php (hunk 28 above)
Comment (barobba) [reply tests/Unit/Models/StateTest.php:3]: Since it was commented out. I think it was broken. I thought you wanted me to delete anything that wasn't implemented. Currently it creates a warning in testing that there are skipped tests.

I can bring it back, though. Would you like that?

File: tests/Unit/Helpers/SessionYearTest.php (hunk 29 above)
Comment (barobba) [reply tests/Unit/Helpers/SessionYearTest.php:31]: Done

File: app/Helpers/Models/LegiScan/SessionYear.php (hunk 26 above)
Comment (barobba) [reply app/Helpers/Models/LegiScan/SessionYear.php:3]: I'm going to just set it per your initial comment.

Comment (barobba) [reply app/Helpers/Models/LegiScan/SessionYear.php:3]: Done

File: app/Models/LegiScan/BillHistory.php (hunk 34 above)
[the remark this answers is not in the document]
Comment (PeterDKC) [reply app/Models/LegiScan/BillHistory.php:81]: That's not a reasonable thought.

[the remark this answers is not in the document]
Comment (PeterDKC) [reply app/Models/LegiScan/BillHistory.php:81]: More specifically it's part of this formalized Code Smell, which you seem to suffer from in a big way.

https://refactoring.guru/smells/speculative-generality

File: app/Models/LegiScan/Progress.php (hunk 35 above)
[the remark this answers is not in the document]
Comment (PeterDKC) [reply app/Models/LegiScan/Progress.php:16]: You chose to put models on top of their schemas. 🤷 Not sure what the objection is at this point.

https://refactoring.guru/smells/speculative-generality

[the remark this answers is not in the document]
Comment (PeterDKC) [reply app/Models/LegiScan/Progress.php:16]: Either fix it or leave it. This is an unhealthy halfway point.

File: tests/Unit/Models/StateTest.php (hunk 28 above)
Comment (PeterDKC) [reply tests/Unit/Models/StateTest.php:3]: I have no recollection of why it would've been skipped, unless you're saying you skipped it. I don't comment out test code.

I guess delete it, I'm unclear why you couldn't just fix the test to match the changes. Doesn't seem hard.

File: app/Models/LegiScan/BillHistory.php (hunk 33 above)
[the remark this answers is not in the document]
Comment (PeterDKC) [reply app/Models/LegiScan/BillHistory.php:69]: Yeah but delete the comment.

STOP LEAVING COMMENTS. rofl

[the remark this answers is not in the document]
Comment (barobba) [reply app/Models/LegiScan/BillHistory.php:69]: Oh yeah. Done!

File: tests/Unit/Models/StateTest.php (hunk 28 above)
Comment (barobba) [reply tests/Unit/Models/StateTest.php]: I was able to add it back and fix it.

**Review by PeterDKC — APPROVED** [review]
